### PR TITLE
[Feature Request] add k3d provider & change --tags to stringSlice

### PR DESCRIPTION
**Pull request by Jason-ZW**
#303 

## Features Requests
1. add k3d provider.

## Bug Fixes
1. `kubeconfig` file should delete associate auth info when the cluster is deleted or rollback logic is trigger.

## Reconstruct
1. make code more strict pattern.
2. remove useless code.
3. make `base.go` more compatible.
4. make the dialer more compatible.
5. change `--tags` to StringSlice e.g. change `--tags a=b,b=c` to `--tags a=b --tags b=c`.

## Limitations
1. highly recommended that K3d run in a Linux / UNIX environment. (MacOS-Docker VM bind-ports problem)

### CODE REVIEW

File: pkg/providers/k3d/k3d.go [hunk 1]
```diff
@@ -0,0 +1,796 @@
...
+	if err != nil {
+		return err
+	}
+	targetOption := reflect.ValueOf(opt).Elem()
+	utils.MergeConfig(sourceOption, targetOption)
+
+	return nil
+}
+
+func (p *K3d) Rollback() error {
```
Comment (JacieChao) [inline pkg/providers/k3d/k3d.go:187]: We have removed the provider interface `Rollback`, so remove the function cause no one is using it.

File: pkg/hosts/websocket.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+	}
+
+	r := NewTerminalReader(d.conn)
+	r.SetResizeFunction(d.ChangeWindowSize)
+	w := NewBinaryWriter(d.conn)
+
+	d.reader = r
+
+	switch kind {
+	case SSHKind:
```
Comment (JacieChao) [inline pkg/hosts/websocket.go:51]: I think it's better to define interface function for the tree kind of dialer to escape from switch case block in each function.

Comment (Jason-ZW) [reply pkg/hosts/websocket.go:51]: Will optimize when I have time, currently not optimize.

**Review by JacieChao — APPROVED** [review]
LGTM

**Pull request merged**